RecoTracker/DisplacedRegionalTracking/plugins/DisplacedRegionProducer.cc

```diff
@@ -0,0 +1,201 @@
+      input_names_(cfg.getParameter<vector<string> >("input_names")),
+      output_names_(cfg.getParameter<vector<string> >("output_names")) {
+  beamSpotToken_ = consumes<reco::BeamSpot>(cfg.getParameter<edm::InputTag>("beamSpot"));
+  trackClustersToken_ =
+      consumes<edm::View<reco::VertexCompositeCandidate> >(cfg.getParameter<edm::InputTag>("trackClusters"));
+
+  unsigned nThreads = cfg.getParameter<unsigned>("nThreads");
+  string singleThreadPool = cfg.getParameter<string>("singleThreadPool");
+  tensorflow::SessionOptions sessionOptions;
+  tensorflow::setThreading(sessionOptions, nThreads, singleThreadPool);
+  session_ = tensorflow::createSession(cache->graphDef, sessionOptions);
+
+  produces<vector<reco::Vertex> >();
+}
+
+DisplacedRegionProducer::~DisplacedRegionProducer() {
+  if (session_ != nullptr)
+    tensorflow::closeSession(session_);
+}
+
+unique_ptr<DisplacedRegionTFCache> DisplacedRegionProducer::initializeGlobalCache(const edm::ParameterSet &cfg) {
+  tensorflow::setLogging("3");
+  string pbFile = cfg.getParameter<edm::FileInPath>("graph_path").fullPath();
+  DisplacedRegionTFCache *cache = new DisplacedRegionTFCache();
+  cache->graphDef = tensorflow::loadGraphDef(pbFile);
+
+  return unique_ptr<DisplacedRegionTFCache>(cache);
+}
+
+void DisplacedRegionProducer::globalEndJob(const DisplacedRegionTFCache *cache) {
+  if (cache->graphDef != nullptr)
+    delete cache->graphDef;
+}
+
+void DisplacedRegionProducer::produce(edm::Event &event, const edm::EventSetup &setup) {
+  edm::Handle<reco::BeamSpot> beamSpot;
+  edm::Handle<edm::View<reco::VertexCompositeCandidate> > trackClusters;
+
+  event.getByToken(beamSpotToken_, beamSpot);
+  const math::XYZVector bs(beamSpot->x0(), beamSpot->y0(), beamSpot->z0());
+  event.getByToken(trackClustersToken_, trackClusters);
+
+  // Initialize distances.
+  list<DisplacedRegion> pseudoROIs;
+  list<Distance> distances;
+  const double minTrackClusterRadius = minRadius_ - rParam_;
+  for (unsigned i = 0; i < trackClusters->size(); i++) {
+    const reco::VertexCompositeCandidate &trackCluster = trackClusters->at(i);
+    const math::XYZVector x(trackCluster.vx(), trackCluster.vy(), trackCluster.vz());
+    if (minRadius_ < 0.0 || minTrackClusterRadius < 0.0 || (x - bs).rho() > minTrackClusterRadius)
+      pseudoROIs.emplace_back(trackClusters, i, rParam_);
+  }
+  if (pseudoROIs.size() > 1) {
+    DisplacedRegionItr secondToLast = pseudoROIs.end();
+    secondToLast--;
+    for (DisplacedRegionItr i = pseudoROIs.begin(); i != secondToLast; i++) {
+      DisplacedRegionItr j = i;
+      j++;
+      for (; j != pseudoROIs.end(); j++)
+        distances.emplace_back(i, j);
+    }
+  }
+
+  // Track clusters farther apart than 4 times rParam_ cannot wind up in the
+  // same ROI, so remove these pairs.
+  const auto pred = [&](const Distance &a) { return a.distance() > 4.0 * rParam_; };
+  distances.remove_if(pred);
```

Comment:
It can be, and I've done this now using `pop_back`. I think this way avoids creating and copying an extra `Distance` object, but please let me know if you see a more efficient way.